lib/client/conntest/ssh.go

```diff
@@ -49,30 +60,212 @@ func NewSSHConnectionTester(cd services.ConnectionsDiagnostic) *SSHConnectionTes
 //   - the SSH Node will append a trace indicating if the has access (RBAC)
 //   - the SSH Node will append a trace indicating if the requested principal is valid for the target Node
 func (s *SSHConnectionTester) TestConnection(ctx context.Context, req TestConnectionRequest) (types.ConnectionDiagnostic, error) {
-	id := uuid.NewString()
-	connectionDiagnostic, err := types.NewConnectionDiagnosticV1(id, map[string]string{},
+	connectionDiagnosticID := uuid.NewString()
+	connectionDiagnostic, err := types.NewConnectionDiagnosticV1(connectionDiagnosticID, map[string]string{},
 		types.ConnectionDiagnosticSpecV1{
-			Message: types.DiagnosticMessageWaiting,
+			Message: types.DiagnosticMessageFailed,
 		})
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	if err := s.cd.CreateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
+	if err := s.clt.CreateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	// TODO(marco): test ssh connection
-	// - create certificate for user with extra field
-	// - create ssh client using that certificate
-	// - if connection fails because of a network error, a trace must be included indicating that the host is not reachable
-	// - other traces will be added by the Node itself (rbac checks, principal)
-	connectionDiagnostic.SetMessage("dry-run")
-	connectionDiagnostic.SetSuccess(true)
+	sshNode, err := s.clt.GetNode(ctx, defaults.Namespace, req.ResourceName)
+	if err != nil {
+		connDiag, err := s.clt.AppendTraceConnectionDiagnostic(ctx, connectionDiagnosticID, types.NewFailedTraceConnectionDiagnostic(
+			uuid.NewString(),
+			types.DiagnosticTraceTypeRBACNode,
+			"Node not found. Ensure the Node exists and your role allows you to access it.",
+			err,
+		))
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		return connDiag, nil
+	}
+
+	_, err = s.clt.AppendTraceConnectionDiagnostic(ctx, connectionDiagnosticID, types.NewSuccessTraceConnectionDiagnostic(
```

Comment:
(This might change due to the concurrency concerns.)

I think we need to keep this as atomic as possible.
There are two places where we add the traces, and having a local copy here would mean we would ignore the added traces in the other place.